src/java/org/apache/cassandra/index/sai/utils/BM25Utils.java

```diff
@@ -186,18 +143,17 @@ public static CloseableIterator<PrimaryKeyWithSortKey> computeScores(CloseableIt
         {
             var tf = docIterator.next();
             documents.add(tf);
-            if (docStats.avgDocLength < 0)
+            if (isOldFormat)
                 totalTermCount += tf.termCount();
         }
 
-        // avgDocLength is unknown since an index format is before {@link Version#ED},
-        // which doesn't store the total term count on disk to read it back.
-        // In such a case the old way of computing avgDocLength is used.
-        if (docStats.avgDocLength < 0)
-        {
-            assert !Version.current().onOrAfter(Version.ED) : "on Version.ED avgDocLength must be known";
-            docStats.avgDocLength = totalTermCount / documents.size();
-        }
+        // An index format before {@link Version#ED} doesn't store the total term count
+        // on the disk to read it back. Thus, for the old format version it is calculated in the old way.
+        double oldAvgDocLength;
+        if (isOldFormat && !documents.isEmpty())
+            oldAvgDocLength = totalTermCount / documents.size();
+        else
+            oldAvgDocLength = 0;
```

Comment:
Did you consider something like the following?

```java
        double avgDocLength = isOldFormat && !documents.isEmpty() ?  totalTermCount / documents.size() : docStats.getAvgDocLength();
```

Reply:
Another approach can be to pass statistics values instead of the document statistics object.